server/impl/meetup.go

```diff
@@ -1,15 +1,762 @@
 package impl
 
 import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-openapi/swag"
+
+	"github.com/go-openapi/strfmt"
+
 	"github.com/go-openapi/runtime/middleware"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
 
+	"go.timothygu.me/downtomeet/server/db"
 	"go.timothygu.me/downtomeet/server/models"
 	"go.timothygu.me/downtomeet/server/restapi/operations"
 )
 
-func GetMeetupID(params operations.GetMeetupIDParams) middleware.Responder {
-	meetup := models.Meetup{
-		ID: models.MeetupID(params.ID),
+/*
+Before running the query in GetMeetup, do
+
+CREATE EXTENSION earthdistance CASCADE;
+
+as a superuser (probably 'postgres')
+*/
+
+/* TODO: remove duplicate meetups */
+func (i *Implementation) GetMeetup(params operations.GetMeetupParams) middleware.Responder {
+	ctx := params.HTTPRequest.Context()
+	logger := log.WithContext(ctx)
+
+	tx := i.DB().WithContext(ctx)
+
+	var tagIds []int
+	var dbTags []db.Tag
+	if err := tx.Where("name IN ?", params.Tags).Find(&dbTags).Error; err != nil {
+		logger.WithError(err).Error("Unable to fetch tags")
+		return InternalServerError{}
+	}
+	for _, tag := range dbTags {
+		tagIds = append(tagIds, int(tag.ID))
+	}
+
+	var meetups []*db.Meetup
+	err := tx.Raw(`
+	SELECT id FROM (
+		SELECT *,
+			earth_distance(ll_to_earth(?, ?),
+			ll_to_earth(location_lat, location_lon)) AS distance_from_me
+		FROM meetups
+	) AS m
+	JOIN meetup_tag AS mt ON m.id = mt.meetup_id
+	WHERE m.distance_from_me < ? AND mt.tag_id IN ?
+	GROUP BY m.id, m.distance_from_me
+	ORDER BY m.distance_from_me
+	LIMIT 100
+	`, params.Lat, params.Lon, params.Radius*1000, tagIds).Scan(&meetups).Error
+
+	if err != nil {
+		logger.WithError(err).Error("Unable to find meetups that fit the given parameters")
+		return InternalServerError{}
+	}
+
+	var idStr string
+	if id := SessionFromContext(ctx).Values[UserID]; id == nil {
+		idStr = ""
+	} else {
```

Comment:
In Go, `idStr` is automatically initialized to "" when you declare it. So we don't need the `if` branch.

```suggestion
	if id := SessionFromContext(ctx).Values[UserID]; id != nil {
```